### PR TITLE
pybind11::args should have been derived from tuple

### PR DESCRIPTION
**Pull request by yungyuc**
args was derived from list, but cpp_function::dispatcher sends a tuple to it->impl (line [#346](https://github.com/pybind/pybind11/blob/1e3be73a521d6c888e09e554b8a239ce28fbf5a8/include/pybind11/pybind11.h#L346) and [#392](https://github.com/pybind/pybind11/blob/1e3be73a521d6c888e09e554b8a239ce28fbf5a8/include/pybind11/pybind11.h#L392) in pybind11.h).  As a result args::size() and args::operator[] don't work at all.  On my mac args::size() returns -1.  Making args a subclass of tuple fixes it.

### CODE REVIEW

File: include/pybind11/pytypes.h [hunk 1]
```diff
@@ -491,7 +491,7 @@ class tuple : public object {
         if (!m_ptr) pybind11_fail("Could not allocate tuple object!");
     }
     size_t size() const { return (size_t) PyTuple_Size(m_ptr); }
-    detail::tuple_accessor operator[](size_t index) const { return detail::tuple_accessor(ptr(), index); }
```
Comment (wjakob) [inline include/pybind11/pytypes.h]: Is this line needed?

**Comment by wjakob** [comment]
Good catch!

File: include/pybind11/pytypes.h (hunk 1 above)
Comment (yungyuc) [inline include/pybind11/pytypes.h]: It can be in another PR if you want.

Comment (wjakob) [inline include/pybind11/pytypes.h]: Having it in the same PR is no problem -- I'm just wondering what the change from ``ptr()`` to ``*this`` is supposed to change here?

Comment (yungyuc) [inline include/pybind11/pytypes.h]: It has no effect to the args issue, but I think it's a missing spot when tuple_accessor constructor changed from taking a PyObject * to handle.  Giving it this->ptr() is the same as *this.  But list::operatorp[]() changed to send *this, so for consistency I think here we should use *this too.

Comment (wjakob) [inline include/pybind11/pytypes.h]: Ok, thanks for the clarification!

Comment (yungyuc) [inline include/pybind11/pytypes.h]: No problem.

An additional note (as a record): [the relevant changeset](https://github.com/yungyuc/pybind11/commit/48548ea4a5735acfb86c67354d3ea6a728169740#diff-0e12743ab8ce3cfb72c5df5d9b956b47R183) shows that both `tuple::tuple()` and `list::list()` changed to take and store `handle`, but only `list::operator[]()` changed to feed `*this`.

Comment (wjakob) [inline include/pybind11/pytypes.h]: Ok, I committed a few more cleanups that replace ``ptr()`` by ``this`` -- let me know if there are still some more that bother you and which I haven't caught.

Comment (yungyuc) [inline include/pybind11/pytypes.h]: They look great.  I confess that I didn't review all the places thoroughly.  The only place I was sure was `tuple::operator[]()`, because I traced around it several times using a debugger.

I am trying to use pybind11 to replace a Cython wrapper for my project.  If I find other places to change I will file new PRs.

Thanks a lot for the great project.
